JDK 25: Cannot introspect nested types from JAR files - incomplete builder generation even with --add-opens

### Summary

  Sundrio cannot introspect nested type chains from JAR files on JDK 25, resulting in incomplete builder generation. This works correctly on JDK 17/21 but fails on JDK 25 due to stricter module encapsulation that cannot be bypassed even with `--add-opens` flags.

  ### Environment

  - **JDK Version**: OpenJDK 25.0.2 (Temurin)
  - **Sundrio Version**: 0.230.1
  - **Maven Version**: 3.9.x
  - **Works on**: JDK 17, JDK 21
  - **Fails on**: JDK 25

  ### The Issue

  When a `@Buildable` class in a consumer module references classes from a dependency JAR:

  1. **Without `--add-opens`**: Builders are not generated at all
  2. **With `--add-opens`**: Builders are generated but incomplete - missing nested collection builder methods

  **Example:**
  - `PolicyBinding` has `@BuildableReference(Selector.class)`
  - `Selector` contains `List<Requirement>` field
  - On JDK 17/21: Sundrio introspects `Selector` from JAR and discovers `Requirement` → generates complete builders ✅
  - On JDK 25: Sundrio reads `Selector` from JAR but CANNOT discover `Requirement` → generates incomplete builders ❌

  **Result**: Missing `addNewMatchExpression()` methods in generated `SelectorFluent` class.

  ### Reproduction

  Minimal reproducible example: https://github.com/ash-thakur-rh/sundrio-jdk25-issue

  **Steps to reproduce:**

  1. **Build on JDK 17/21** (works):
   ```bash
    export JAVA_HOME=/path/to/jdk-17
    mvn clean install
    # ✅ Build succeeds - complete builders generated
   ```

  2. Build on JDK 25 without --add-opens (fails):
  ```bash
  export JAVA_HOME=/path/to/jdk-25
  mvn clean install
  # ❌ Build fails - builders not generated
  # [ERROR] PolicyBindingBuilder not found
  ```

  3. Build on JDK 25 with --add-opens (fails differently):
  ```bash
  export JAVA_HOME=/path/to/jdk-25
  # Uncomment --add-opens compiler args in pom.xml (lines 46-75)
  mvn clean install
  # ❌ Build fails - builders generated but incomplete
  # [ERROR] cannot find symbol: method addNewMatchExpression()
  ```

  Expected Behavior

  On JDK 25, Sundrio should introspect nested types from JAR files just like it does on JDK 17/21, generating complete builders with all nested collection methods.

  Actual Behavior

  On JDK 25, even with --add-opens flags, Sundrio cannot introspect nested types from JAR files:
  - It reads the parent class (Selector) from the JAR
  - It sees the collection field (List<Requirement>)
  - It cannot introspect into the nested type (Requirement) to generate builder methods
  - Results in incomplete builders missing addNew*() methods

  Root Cause

  JDK 25's stricter module encapsulation prevents Sundrio from performing deep reflection into classes loaded from JAR files. The --add-opens flags allow basic builder generation but don't enable the nested type discovery that
  worked on JDK 17/21.

--- a/annotations/builder/src/it/buildable-concrete-collections/src/test/java/io/sundr/it/LinkedListHolderTest.java
+++ b/annotations/builder/src/it/buildable-concrete-collections/src/test/java/io/sundr/it/LinkedListHolderTest.java
@@ -49,4 +49,66 @@ public class LinkedListHolderTest {
         assertEquals(1, item.getList().size());
         assertEquals("foo", item.getList().get(0).getName());
     }
+
+    @Test
+    public void shouldSupportNestedBuilderMethods() {
+        // Test nested builder pattern: addNewList()...endList()
+        // This verifies JDK 25 compatibility for generic type parameter resolution
+        LinkedListHolder item = new LinkedListHolderBuilder()
+            .addNewList()
+                .withId(42)
+                .withName("nested-builder-test")
+            .endList()
+            .build();
+
+        assertNotNull(item.getList());
+        assertEquals(1, item.getList().size());
+        assertEquals(42, item.getList().get(0).getId());
+        assertEquals("nested-builder-test", item.getList().get(0).getName());
+    }
+
+    @Test
+    public void shouldSupportMultipleNestedBuilders() {
+        // Test multiple nested builder calls
+        // This tests that annotation processor handles List<T> bounds correctly
+        LinkedListHolder item = new LinkedListHolderBuilder()
+            .addNewList()
+                .withId(1)
+                .withName("first")
+            .endList()
+            .addNewList()
+                .withId(2)
+                .withName("second")
+            .endList()
+            .build();
+
+        assertNotNull(item.getList());
+        assertEquals(2, item.getList().size());
+        assertEquals(1, item.getList().get(0).getId());
+        assertEquals("first", item.getList().get(0).getName());
+        assertEquals(2, item.getList().get(1).getId());
+        assertEquals("second", item.getList().get(1).getName());
+    }
+
+    @Test
+    public void shouldSupportEditNestedBuilder() {
+        // Test editing existing items with nested builders
+        LinkedListHolder original = new LinkedListHolderBuilder()
+            .addNewList()
+                .withId(10)
+                .withName("original")
+            .endList()
+            .build();
+
+        LinkedListHolder modified = new LinkedListHolderBuilder(original)
+            .editList(0)
+                .withId(20)
+                .withName("modified")
+            .endList()
+            .build();
+
+        assertEquals(1, modified.getList().size());
+        assertEquals(20, modified.getList().get(0).getId());
+        assertEquals("modified", modified.getList().get(0).getName());
+    }
 }

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/pom.xml
@@ -1,0 +1,60 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.sundr.it</groupId>
+        <artifactId>cross-module-buildable-ref</artifactId>
+        <version>@project.version@</version>
+    </parent>
+
+    <artifactId>cross-module-buildable-ref-consumer</artifactId>
+    <name>Sundrio :: Annotations :: Builders :: IT :: Cross-Module BuildableReference :: Consumer</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.sundr.it</groupId>
+            <artifactId>cross-module-buildable-ref-library</artifactId>
+            <version>@project.version@</version>
+        </dependency>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>sundr-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>builder-annotations</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <scope>test</scope>
+        </dependency>
+    </dependencies>
+
+    <build>
+        <plugins>
+            <!--
+                Use annotationProcessorPaths to isolate the annotation processor ClassLoader
+                from the compile classpath. This reproduces the cross-module @BuildableReference
+                scenario: the library JAR (with MatchExpressionBuilder) is only on the compile
+                classpath, not the processor path. Without the fix in BuilderUtils.isBuildable(TypeRef),
+                the processor cannot detect that MatchExpression is buildable and omits
+                addNewMatchExpression() from the generated LabelSelectorFluent.
+            -->
+            <plugin>
+                <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <configuration>
+                    <annotationProcessorPaths>
+                        <path>
+                            <groupId>io.sundr</groupId>
+                            <artifactId>builder-annotations</artifactId>
+                            <version>${sundr.version}</version>
+                        </path>
+                    </annotationProcessorPaths>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+</project>

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/pom.xml
@@ -26,8 +26,8 @@
             <artifactId>builder-annotations</artifactId>
         </dependency>
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
             <scope>test</scope>
         </dependency>
     </dependencies>

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/main/java/io/sundr/it/PolicyBinding.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/main/java/io/sundr/it/PolicyBinding.java
@@ -26,7 +26,7 @@ import io.sundr.builder.annotations.BuildableReference;
  * methods (addNewMatchExpression()) on LabelSelectorFluent even though MatchExpression
  * comes from a compiled JAR rather than source.
  */
-@Buildable(refs = { @BuildableReference(LabelSelector.class) })
+@Buildable(editableEnabled = false, validationEnabled = false, generateBuilderPackage = true, lazyCollectionInitEnabled = false, builderPackage = "io.sundrio.it", refs = { @BuildableReference(LabelSelector.class) })
 public class PolicyBinding {
 
     private final String name;

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/main/java/io/sundr/it/PolicyBinding.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/main/java/io/sundr/it/PolicyBinding.java
@@ -1,0 +1,47 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package io.sundr.it;
+
+import io.sundr.builder.annotations.Buildable;
+import io.sundr.builder.annotations.BuildableReference;
+
+/**
+ * Consumer class that references LabelSelector from the library dependency via @BuildableReference.
+ * This tests the cross-module scenario where LabelSelector has a List&lt;MatchExpression&gt; field
+ * and MatchExpression is also buildable. The consumer module should generate nested builder
+ * methods (addNewMatchExpression()) on LabelSelectorFluent even though MatchExpression
+ * comes from a compiled JAR rather than source.
+ */
+@Buildable(refs = { @BuildableReference(LabelSelector.class) })
+public class PolicyBinding {
+
+    private final String name;
+    private final LabelSelector selector;
+
+    public PolicyBinding(String name, LabelSelector selector) {
+        this.name = name;
+        this.selector = selector;
+    }
+
+    public String getName() {
+        return name;
+    }
+
+    public LabelSelector getSelector() {
+        return selector;
+    }
+}

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/test/java/io/sundr/it/CrossModuleBuildableRefTest.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/test/java/io/sundr/it/CrossModuleBuildableRefTest.java
@@ -1,0 +1,85 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+package io.sundr.it;
+
+import org.junit.Test;
+
+import java.util.Arrays;
+
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertNotNull;
+
+/**
+ * Verifies that nested builder methods are generated correctly when a @BuildableReference
+ * type (LabelSelector) has fields whose element types (MatchExpression) are also buildable
+ * and come from a compiled dependency JAR rather than source.
+ *
+ * Without the fix, LabelSelectorFluent would be missing addNewMatchExpression() and the
+ * MatchExpressionsNested inner class, because MatchExpression would not be registered
+ * in the buildable repository when processing the consumer module.
+ */
+public class CrossModuleBuildableRefTest {
+
+    @Test
+    public void testNestedBuilderViaBuilableReference() {
+        // LabelSelector comes from the library JAR via @BuildableReference.
+        // MatchExpression is a field element type that must also be recognized as buildable
+        // so that addNewMatchExpression() is generated on LabelSelectorFluent.
+        PolicyBinding binding = new PolicyBindingBuilder()
+            .withName("test-binding")
+            .withNewSelector()
+                .addNewMatchExpression()
+                    .withKey("environment")
+                    .withOperator("In")
+                    .withValues(Arrays.asList("production", "staging"))
+                .endMatchExpression()
+            .endSelector()
+            .build();
+
+        assertNotNull(binding);
+        assertEquals("test-binding", binding.getName());
+        assertNotNull(binding.getSelector());
+        assertEquals(1, binding.getSelector().getMatchExpressions().size());
+
+        MatchExpression expr = binding.getSelector().getMatchExpressions().get(0);
+        assertEquals("environment", expr.getKey());
+        assertEquals("In", expr.getOperator());
+        assertEquals(2, expr.getValues().size());
+    }
+
+    @Test
+    public void testMultipleNestedExpressionsViaBuilableReference() {
+        PolicyBinding binding = new PolicyBindingBuilder()
+            .withName("multi")
+            .withNewSelector()
+                .addNewMatchExpression()
+                    .withKey("app")
+                    .withOperator("In")
+                    .withValues(Arrays.asList("frontend"))
+                .endMatchExpression()
+                .addNewMatchExpression()
+                    .withKey("tier")
+                    .withOperator("NotIn")
+                    .withValues(Arrays.asList("cache"))
+                .endMatchExpression()
+            .endSelector()
+            .build();
+
+        assertEquals(2, binding.getSelector().getMatchExpressions().size());
+        assertEquals("app", binding.getSelector().getMatchExpressions().get(0).getKey());
+        assertEquals("tier", binding.getSelector().getMatchExpressions().get(1).getKey());
+    }
+}

--- a/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/test/java/io/sundr/it/CrossModuleBuildableRefTest.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/consumer/src/test/java/io/sundr/it/CrossModuleBuildableRefTest.java
@@ -15,12 +15,12 @@
  */
 package io.sundr.it;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
 
 import java.util.Arrays;
 
-import static org.junit.Assert.assertEquals;
-import static org.junit.Assert.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
 
 /**
  * Verifies that nested builder methods are generated correctly when a @BuildableReference

--- a/annotations/builder/src/it/cross-module-buildable-ref/library/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/library/pom.xml
@@ -1,0 +1,24 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <parent>
+        <groupId>io.sundr.it</groupId>
+        <artifactId>cross-module-buildable-ref</artifactId>
+        <version>@project.version@</version>
+    </parent>
+
+    <artifactId>cross-module-buildable-ref-library</artifactId>
+    <name>Sundrio :: Annotations :: Builders :: IT :: Cross-Module BuildableReference :: Library</name>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>sundr-core</artifactId>
+        </dependency>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>builder-annotations</artifactId>
+        </dependency>
+    </dependencies>
+</project>

--- a/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/LabelSelector.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/LabelSelector.java
@@ -20,7 +20,7 @@ import io.sundr.builder.annotations.Buildable;
 import java.util.List;
 import java.util.Map;
 
-@Buildable
+@Buildable(editableEnabled = false, validationEnabled = false, generateBuilderPackage = true, lazyCollectionInitEnabled = false, builderPackage = "io.sundrio.it")
 public class LabelSelector {
 
     private final Map<String, String> matchLabels;

--- a/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/LabelSelector.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/LabelSelector.java
@@ -1,0 +1,41 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package io.sundr.it;
+
+import io.sundr.builder.annotations.Buildable;
+import java.util.List;
+import java.util.Map;
+
+@Buildable
+public class LabelSelector {
+
+    private final Map<String, String> matchLabels;
+    private final List<MatchExpression> matchExpressions;
+
+    public LabelSelector(Map<String, String> matchLabels, List<MatchExpression> matchExpressions) {
+        this.matchLabels = matchLabels;
+        this.matchExpressions = matchExpressions;
+    }
+
+    public Map<String, String> getMatchLabels() {
+        return matchLabels;
+    }
+
+    public List<MatchExpression> getMatchExpressions() {
+        return matchExpressions;
+    }
+}

--- a/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/MatchExpression.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/MatchExpression.java
@@ -19,7 +19,7 @@ package io.sundr.it;
 import io.sundr.builder.annotations.Buildable;
 import java.util.List;
 
-@Buildable
+@Buildable(editableEnabled = false, validationEnabled = false, generateBuilderPackage = true, lazyCollectionInitEnabled = false, builderPackage = "io.sundrio.it")
 public class MatchExpression {
 
     private final String key;

--- a/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/MatchExpression.java
+++ b/annotations/builder/src/it/cross-module-buildable-ref/library/src/main/java/io/sundr/it/MatchExpression.java
@@ -1,0 +1,46 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package io.sundr.it;
+
+import io.sundr.builder.annotations.Buildable;
+import java.util.List;
+
+@Buildable
+public class MatchExpression {
+
+    private final String key;
+    private final String operator;
+    private final List<String> values;
+
+    public MatchExpression(String key, String operator, List<String> values) {
+        this.key = key;
+        this.operator = operator;
+        this.values = values;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    public String getOperator() {
+        return operator;
+    }
+
+    public List<String> getValues() {
+        return values;
+    }
+}

--- a/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
@@ -1,0 +1,75 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+       Copyright 2026 The original authors.
+
+       Licensed under the Apache License, Version 2.0 (the "License");
+       you may not use this file except in compliance with the License.
+       You may obtain a copy of the License at
+
+           http://www.apache.org/licenses/LICENSE-2.0
+
+       Unless required by applicable law or agreed to in writing, software
+       distributed under the License is distributed on an "AS IS" BASIS,
+       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+       See the License for the specific language governing permissions and
+       limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.sundr.it</groupId>
+    <artifactId>cross-module-buildable-ref</artifactId>
+    <version>@project.version@</version>
+    <packaging>pom</packaging>
+    <name>Sundrio :: Annotations :: Builders :: Integration Tests :: Cross-Module BuildableReference</name>
+
+    <modules>
+        <module>library</module>
+        <module>consumer</module>
+    </modules>
+
+    <properties>
+        <sundr.version>@project.version@</sundr.version>
+        <junit.version>@junit.version@</junit.version>
+    </properties>
+
+    <build>
+        <pluginManagement>
+            <plugins>
+                <plugin>
+                    <groupId>org.apache.maven.plugins</groupId>
+                    <artifactId>maven-compiler-plugin</artifactId>
+                    <version>3.14.0</version>
+                    <configuration>
+                        <source>11</source>
+                        <target>11</target>
+                        <compilerArgs>
+                            <arg>-proc:full</arg>
+                        </compilerArgs>
+                    </configuration>
+                </plugin>
+            </plugins>
+        </pluginManagement>
+    </build>
+
+    <dependencyManagement>
+        <dependencies>
+            <dependency>
+                <groupId>io.sundr</groupId>
+                <artifactId>sundr-core</artifactId>
+                <version>${sundr.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>io.sundr</groupId>
+                <artifactId>builder-annotations</artifactId>
+                <version>${sundr.version}</version>
+            </dependency>
+            <dependency>
+                <groupId>junit</groupId>
+                <artifactId>junit</artifactId>
+                <version>${junit.version}</version>
+            </dependency>
+        </dependencies>
+    </dependencyManagement>
+</project>

--- a/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
@@ -42,8 +42,8 @@
                     <artifactId>maven-compiler-plugin</artifactId>
                     <version>3.14.0</version>
                     <configuration>
-                        <source>11</source>
-                        <target>11</target>
+                        <source>${java.specification.version}</source>
+                        <target>${java.specification.version}</target>
                         <compilerArgs>
                             <arg>-proc:full</arg>
                         </compilerArgs>

--- a/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
+++ b/annotations/builder/src/it/cross-module-buildable-ref/pom.xml
@@ -31,7 +31,7 @@
 
     <properties>
         <sundr.version>@project.version@</sundr.version>
-        <junit.version>@junit.version@</junit.version>
+        <junit.version>@junit-jupiter.version@</junit.version>
     </properties>
 
     <build>
@@ -66,8 +66,8 @@
                 <version>${sundr.version}</version>
             </dependency>
             <dependency>
-                <groupId>junit</groupId>
-                <artifactId>junit</artifactId>
+                <groupId>org.junit.jupiter</groupId>
+                <artifactId>junit-jupiter-api</artifactId>
                 <version>${junit.version}</version>
             </dependency>
         </dependencies>

--- a/annotations/builder/src/it/nested-builders/pom.xml
+++ b/annotations/builder/src/it/nested-builders/pom.xml
@@ -30,8 +30,8 @@
                 <artifactId>maven-compiler-plugin</artifactId>
                 <version>3.14.0</version>
                 <configuration>
-                    <source>11</source>
-                    <target>11</target>
+                    <source>${java.specification.version}</source>
+                    <target>${java.specification.version}</target>
                     <compilerArgs>
                         <arg>-proc:full</arg>
                     </compilerArgs>

--- a/annotations/builder/src/it/nested-builders/pom.xml
+++ b/annotations/builder/src/it/nested-builders/pom.xml
@@ -53,9 +53,9 @@
         </dependency>
 
         <dependency>
-            <groupId>junit</groupId>
-            <artifactId>junit</artifactId>
-            <version>@junit.version@</version>
+            <groupId>org.junit.jupiter</groupId>
+            <artifactId>junit-jupiter-api</artifactId>
+            <version>@junit-jupiter.version@</version>
         </dependency>
     </dependencies>
 </project>

--- a/annotations/builder/src/it/nested-builders/pom.xml
+++ b/annotations/builder/src/it/nested-builders/pom.xml
@@ -1,0 +1,61 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!--
+       Copyright 2026 The original authors.
+
+       Licensed under the Apache License, Version 2.0 (the "License");
+       you may not use this file except in compliance with the License.
+       You may obtain a copy of the License at
+
+           http://www.apache.org/licenses/LICENSE-2.0
+
+       Unless required by applicable law or agreed to in writing, software
+       distributed under the License is distributed on an "AS IS" BASIS,
+       WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+       See the License for the specific language governing permissions and
+       limitations under the License.
+  -->
+
+<project xmlns="http://maven.apache.org/POM/4.0.0" xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance" xsi:schemaLocation="http://maven.apache.org/POM/4.0.0 http://maven.apache.org/xsd/maven-4.0.0.xsd">
+    <modelVersion>4.0.0</modelVersion>
+
+    <groupId>io.sundr.it</groupId>
+    <artifactId>nested-builders</artifactId>
+    <version>@project.version@</version>
+    <name>Sundrio :: Annotations :: Builders :: Integration Tests :: Nested Builders</name>
+
+    <build>
+        <plugins>
+            <plugin>
+              <groupId>org.apache.maven.plugins</groupId>
+                <artifactId>maven-compiler-plugin</artifactId>
+                <version>3.14.0</version>
+                <configuration>
+                    <source>11</source>
+                    <target>11</target>
+                    <compilerArgs>
+                        <arg>-proc:full</arg>
+                    </compilerArgs>
+                </configuration>
+            </plugin>
+        </plugins>
+    </build>
+
+    <dependencies>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>sundr-core</artifactId>
+            <version>@project.version@</version>
+        </dependency>
+        <dependency>
+            <groupId>io.sundr</groupId>
+            <artifactId>builder-annotations</artifactId>
+            <version>@project.version@</version>
+        </dependency>
+
+        <dependency>
+            <groupId>junit</groupId>
+            <artifactId>junit</artifactId>
+            <version>@junit.version@</version>
+        </dependency>
+    </dependencies>
+</project>

--- a/annotations/builder/src/it/nested-builders/src/main/java/io/sundr/it/LabelSelector.java
+++ b/annotations/builder/src/it/nested-builders/src/main/java/io/sundr/it/LabelSelector.java
@@ -1,0 +1,41 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package io.sundr.it;
+
+import io.sundr.builder.annotations.Buildable;
+import java.util.List;
+import java.util.Map;
+
+@Buildable
+public class LabelSelector {
+
+    private final Map<String, String> matchLabels;
+    private final List<MatchExpression> matchExpressions;
+
+    public LabelSelector(Map<String, String> matchLabels, List<MatchExpression> matchExpressions) {
+        this.matchLabels = matchLabels;
+        this.matchExpressions = matchExpressions;
+    }
+
+    public Map<String, String> getMatchLabels() {
+        return matchLabels;
+    }
+
+    public List<MatchExpression> getMatchExpressions() {
+        return matchExpressions;
+    }
+}

--- a/annotations/builder/src/it/nested-builders/src/main/java/io/sundr/it/MatchExpression.java
+++ b/annotations/builder/src/it/nested-builders/src/main/java/io/sundr/it/MatchExpression.java
@@ -1,0 +1,46 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+
+package io.sundr.it;
+
+import io.sundr.builder.annotations.Buildable;
+import java.util.List;
+
+@Buildable
+public class MatchExpression {
+
+    private final String key;
+    private final String operator;
+    private final List<String> values;
+
+    public MatchExpression(String key, String operator, List<String> values) {
+        this.key = key;
+        this.operator = operator;
+        this.values = values;
+    }
+
+    public String getKey() {
+        return key;
+    }
+
+    public String getOperator() {
+        return operator;
+    }
+
+    public List<String> getValues() {
+        return values;
+    }
+}

--- a/annotations/builder/src/it/nested-builders/src/test/java/io/sundr/it/NestedBuilderTest.java
+++ b/annotations/builder/src/it/nested-builders/src/test/java/io/sundr/it/NestedBuilderTest.java
@@ -1,0 +1,171 @@
+/*
+ *      Copyright 2026 The original authors.
+ *
+ *      Licensed under the Apache License, Version 2.0 (the "License");
+ *      you may not use this file except in compliance with the License.
+ *      You may obtain a copy of the License at
+ *
+ *          http://www.apache.org/licenses/LICENSE-2.0
+ *
+ *      Unless required by applicable law or agreed to in writing, software
+ *      distributed under the License is distributed on an "AS IS" BASIS,
+ *      WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ *      See the License for the specific language governing permissions and
+ *      limitations under the License.
+ */
+package io.sundr.it;
+
+import org.junit.Test;
+import java.util.Arrays;
+
+import static org.junit.Assert.*;
+
+/**
+ * Test nested builder methods for generic types.
+ * This verifies that annotation processing correctly generates
+ * addNew* and end* methods for buildable collection elements.
+ */
+public class NestedBuilderTest {
+
+    @Test
+    public void testAddNewMatchExpression() {
+        // Test the nested builder pattern: addNewMatchExpression()...endMatchExpression()
+        LabelSelector selector = new LabelSelectorBuilder()
+            .addNewMatchExpression()
+                .withKey("app")
+                .withOperator("In")
+                .withValues(Arrays.asList("frontend", "backend"))
+            .endMatchExpression()
+            .build();
+
+        assertNotNull(selector);
+        assertNotNull(selector.getMatchExpressions());
+        assertEquals(1, selector.getMatchExpressions().size());
+
+        MatchExpression expr = selector.getMatchExpressions().get(0);
+        assertEquals("app", expr.getKey());
+        assertEquals("In", expr.getOperator());
+        assertEquals(2, expr.getValues().size());
+        assertTrue(expr.getValues().contains("frontend"));
+        assertTrue(expr.getValues().contains("backend"));
+    }
+
+    @Test
+    public void testMultipleNestedBuilders() {
+        // Test multiple nested builders
+        LabelSelector selector = new LabelSelectorBuilder()
+            .addNewMatchExpression()
+                .withKey("environment")
+                .withOperator("In")
+                .withValues(Arrays.asList("production"))
+            .endMatchExpression()
+            .addNewMatchExpression()
+                .withKey("tier")
+                .withOperator("NotIn")
+                .withValues(Arrays.asList("cache"))
+            .endMatchExpression()
+            .build();
+
+        assertNotNull(selector);
+        assertNotNull(selector.getMatchExpressions());
+        assertEquals(2, selector.getMatchExpressions().size());
+
+        MatchExpression expr1 = selector.getMatchExpressions().get(0);
+        assertEquals("environment", expr1.getKey());
+        assertEquals("In", expr1.getOperator());
+
+        MatchExpression expr2 = selector.getMatchExpressions().get(1);
+        assertEquals("tier", expr2.getKey());
+        assertEquals("NotIn", expr2.getOperator());
+    }
+
+    @Test
+    public void testMixedAddMethods() {
+        // Test mixing addNew* with addTo* methods
+        MatchExpression existingExpr = new MatchExpressionBuilder()
+            .withKey("existing")
+            .withOperator("Exists")
+            .build();
+
+        LabelSelector selector = new LabelSelectorBuilder()
+            .addToMatchExpressions(existingExpr)
+            .addNewMatchExpression()
+                .withKey("new")
+                .withOperator("In")
+                .withValues(Arrays.asList("value"))
+            .endMatchExpression()
+            .build();
+
+        assertNotNull(selector);
+        assertEquals(2, selector.getMatchExpressions().size());
+        assertEquals("existing", selector.getMatchExpressions().get(0).getKey());
+        assertEquals("new", selector.getMatchExpressions().get(1).getKey());
+    }
+
+    @Test
+    public void testNestedBuilderWithEmptyValues() {
+        // Test nested builder with null/empty values
+        LabelSelector selector = new LabelSelectorBuilder()
+            .addNewMatchExpression()
+                .withKey("test")
+                .withOperator("Exists")
+            .endMatchExpression()
+            .build();
+
+        assertNotNull(selector);
+        assertEquals(1, selector.getMatchExpressions().size());
+        assertEquals("test", selector.getMatchExpressions().get(0).getKey());
+        assertEquals("Exists", selector.getMatchExpressions().get(0).getOperator());
+    }
+
+    @Test
+    public void testEditMatchExpression() {
+        // Test editing an existing match expression
+        LabelSelector original = new LabelSelectorBuilder()
+            .addNewMatchExpression()
+                .withKey("original")
+                .withOperator("In")
+                .withValues(Arrays.asList("v1"))
+            .endMatchExpression()
+            .build();
+
+        LabelSelector modified = new LabelSelectorBuilder(original)
+            .editMatchExpression(0)
+                .withKey("modified")
+                .withOperator("NotIn")
+                .withValues(Arrays.asList("v2", "v3"))
+            .endMatchExpression()
+            .build();
+
+        assertEquals(1, modified.getMatchExpressions().size());
+        assertEquals("modified", modified.getMatchExpressions().get(0).getKey());
+        assertEquals("NotIn", modified.getMatchExpressions().get(0).getOperator());
+        assertEquals(2, modified.getMatchExpressions().get(0).getValues().size());
+    }
+
+    @Test
+    public void testComplexNestedStructure() {
+        // Test a complex nested structure combining multiple features
+        LabelSelector selector = new LabelSelectorBuilder()
+            .addToMatchLabels("version", "1.0")
+            .addToMatchLabels("release", "stable")
+            .addNewMatchExpression()
+                .withKey("component")
+                .withOperator("In")
+                .withValues(Arrays.asList("api", "web", "worker"))
+            .endMatchExpression()
+            .addNewMatchExpression()
+                .withKey("deprecated")
+                .withOperator("DoesNotExist")
+            .endMatchExpression()
+            .build();
+
+        assertNotNull(selector.getMatchLabels());
+        assertEquals(2, selector.getMatchLabels().size());
+        assertEquals("1.0", selector.getMatchLabels().get("version"));
+        assertEquals("stable", selector.getMatchLabels().get("release"));
+
+        assertEquals(2, selector.getMatchExpressions().size());
+        assertEquals(3, selector.getMatchExpressions().get(0).getValues().size());
+    }
+}

--- a/annotations/builder/src/it/nested-builders/src/test/java/io/sundr/it/NestedBuilderTest.java
+++ b/annotations/builder/src/it/nested-builders/src/test/java/io/sundr/it/NestedBuilderTest.java
@@ -15,10 +15,12 @@
  */
 package io.sundr.it;
 
-import org.junit.Test;
+import org.junit.jupiter.api.Test;
 import java.util.Arrays;
 
-import static org.junit.Assert.*;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotNull;
+import static org.junit.jupiter.api.Assertions.assertTrue;
 
 /**
  * Test nested builder methods for generic types.

--- a/annotations/builder/src/main/java/io/sundr/builder/internal/utils/BuilderUtils.java
+++ b/annotations/builder/src/main/java/io/sundr/builder/internal/utils/BuilderUtils.java
@@ -47,8 +47,10 @@ import java.util.stream.Collectors;
 import java.util.stream.Stream;
 
 import javax.lang.model.element.Element;
+import javax.lang.model.element.ElementKind;
 import javax.lang.model.element.TypeElement;
 import javax.lang.model.type.MirroredTypeException;
+import javax.lang.model.util.Elements;
 
 import io.sundr.adapter.api.AdapterContext;
 import io.sundr.adapter.api.Adapters;
@@ -151,7 +153,28 @@ public class BuilderUtils {
   }
 
   public static boolean isBuildable(TypeRef typeRef) {
-    return isRegisteredAsBuildable(typeRef) && canBeBuilt(typeRef);
+    if (isRegisteredAsBuildable(typeRef) && canBeBuilt(typeRef)) {
+      return true;
+    }
+    if (!(typeRef instanceof ClassRef)) {
+      return false;
+    }
+    // Fallback for cross-module @BuildableReference: check if a sundrio-generated *Builder exists
+    // on the APT classpath. @Buildable has SOURCE retention so compiled JARs don't carry it;
+    // using getTypeElement() is the only JDK-version-agnostic way to detect a pre-compiled builder.
+    // We verify it's a sundrio builder (not e.g. java.lang.StringBuilder or ProcessBuilder) by
+    // checking it declares a build() method.
+    String builderFQCN = ((ClassRef) typeRef).getFullyQualifiedName() + "Builder";
+    Elements elements = BuilderContextManager.getContext().getElements();
+    if (elements == null) {
+      return false;
+    }
+    TypeElement builderElement = elements.getTypeElement(builderFQCN);
+    if (builderElement == null) {
+      return false;
+    }
+    return builderElement.getEnclosedElements().stream()
+        .anyMatch(e -> e.getKind() == ElementKind.METHOD && e.getSimpleName().contentEquals("build"));
   }
 
   public static boolean isRegisteredAsBuildable(TypeDef typeDef) {
@@ -176,7 +199,13 @@ public class BuilderUtils {
     String builderFQCN = ref.getFullyQualifiedName() + "Builder";
     TypeDef builder = BuilderContextManager.getContext().getDefinitionRepository().getDefinition(builderFQCN);
     if (builder == null) {
-      return false;
+      Elements elements = BuilderContextManager.getContext().getElements();
+      if (elements == null) {
+        return false;
+      }
+      TypeElement builderElement = elements.getTypeElement(builderFQCN);
+      return builderElement != null && builderElement.getEnclosedElements().stream()
+          .anyMatch(e -> e.getKind() == ElementKind.METHOD && e.getSimpleName().contentEquals("build"));
     }
 
     return builder.getMethods()

--- a/annotations/builder/src/main/java/io/sundr/builder/internal/utils/BuilderUtils.java
+++ b/annotations/builder/src/main/java/io/sundr/builder/internal/utils/BuilderUtils.java
@@ -153,28 +153,10 @@ public class BuilderUtils {
   }
 
   public static boolean isBuildable(TypeRef typeRef) {
-    if (isRegisteredAsBuildable(typeRef) && canBeBuilt(typeRef)) {
-      return true;
-    }
     if (!(typeRef instanceof ClassRef)) {
       return false;
     }
-    // Fallback for cross-module @BuildableReference: check if a sundrio-generated *Builder exists
-    // on the APT classpath. @Buildable has SOURCE retention so compiled JARs don't carry it;
-    // using getTypeElement() is the only JDK-version-agnostic way to detect a pre-compiled builder.
-    // We verify it's a sundrio builder (not e.g. java.lang.StringBuilder or ProcessBuilder) by
-    // checking it declares a build() method.
-    String builderFQCN = ((ClassRef) typeRef).getFullyQualifiedName() + "Builder";
-    Elements elements = BuilderContextManager.getContext().getElements();
-    if (elements == null) {
-      return false;
-    }
-    TypeElement builderElement = elements.getTypeElement(builderFQCN);
-    if (builderElement == null) {
-      return false;
-    }
-    return builderElement.getEnclosedElements().stream()
-        .anyMatch(e -> e.getKind() == ElementKind.METHOD && e.getSimpleName().contentEquals("build"));
+    return (isRegisteredAsBuildable(typeRef) && canBeBuilt(typeRef)) || hasBuilder((ClassRef) typeRef);
   }
 
   public static boolean isRegisteredAsBuildable(TypeDef typeDef) {
@@ -192,29 +174,39 @@ public class BuilderUtils {
    * @return True if buildable repository contains the ref or builder for the reference is present.
    */
   public static boolean isBuildable(ClassRef ref) {
-    if (BuilderContextManager.getContext().getBuildableRepository().isBuildable(ref)) {
-      return true;
-    }
+    return BuilderContextManager.getContext().getBuildableRepository().isBuildable(ref) || hasBuilder(ref);
+  }
 
+  /**
+   * Checks if a sundrio-generated Builder class exists for the given {@link ClassRef}.
+   *
+   * Looks first in the definition repository (same-module types), then falls back to the APT
+   * {@link Elements} API for cross-module types whose {@code @Buildable} annotation is not visible
+   * at compile time (SOURCE retention). A class is considered to have a builder if a type named
+   * {@code <FQCN>Builder} exists and declares a {@code build()} method.
+   *
+   * @param ref The class reference to check.
+   * @return True if a matching builder is found.
+   */
+  public static boolean hasBuilder(ClassRef ref) {
     String builderFQCN = ref.getFullyQualifiedName() + "Builder";
     TypeDef builder = BuilderContextManager.getContext().getDefinitionRepository().getDefinition(builderFQCN);
-    if (builder == null) {
-      Elements elements = BuilderContextManager.getContext().getElements();
-      if (elements == null) {
-        return false;
-      }
-      TypeElement builderElement = elements.getTypeElement(builderFQCN);
-      return builderElement != null && builderElement.getEnclosedElements().stream()
-          .anyMatch(e -> e.getKind() == ElementKind.METHOD && e.getSimpleName().contentEquals("build"));
+    if (builder != null) {
+      return builder.getMethods()
+          .stream()
+          .filter(m -> "build".equals(m.getName()))
+          .filter(m -> m.getReturnType() instanceof ClassRef)
+          .map(m -> (ClassRef) m.getReturnType())
+          .anyMatch(r -> Assignable.isAssignable(r).from(ref));
     }
 
-    return builder.getMethods()
-        .stream()
-        .filter(m -> "build".equals(m.getName()))
-        .filter(m -> m.getReturnType() instanceof ClassRef)
-        .map(m -> (ClassRef) m.getReturnType())
-        .filter(r -> Assignable.isAssignable(r).from(ref))
-        .count() > 0;
+    Elements elements = BuilderContextManager.getContext().getElements();
+    if (elements == null) {
+      return false;
+    }
+    TypeElement builderElement = elements.getTypeElement(builderFQCN);
+    return builderElement != null && builderElement.getEnclosedElements().stream()
+        .anyMatch(e -> e.getKind() == ElementKind.METHOD && e.getSimpleName().contentEquals("build"));
   }
 
   public static boolean isOrHasBuildableDescendants(Field field) {

